### PR TITLE
cpu/nrf51822/periph/random Power optimization

### PR DESCRIPTION
**Pull request by d00616**
My goal is to drive one NRF51 controller via CR2032 cell for an minimum of three years with radio network. This needs to reduce average current to 7.5uA.  This is one small step to reach this goal.

This PR has some effects:
1. PAN #21 fix for better RNG quality
2. random_read doesn't hang when random is powered off
3. auto powerdown RNG to save 16uA current
4. reduce code size for RNG power handling
5. fix hanging random after poweroff() and poweron()

### CODE REVIEW

**Comment by d00616** [comment]
I have changed this driver completely. I have measured an reduced power consumption of up to 6.27mA .

**Comment by haukepetersen** [comment]
I think the general idea is good, but two remarks:

1. Your code does not quite work: if there is an interrupt from another source while you are sleeping, you will read a wrong/old value into your results
2. Why are you moving the power-ctrl into the read function instead of calling them explicitly? This kind of breaks the concept of the periph API...

**Comment by d00616** [comment]
I think the general idea is good, but two remarks:
>
>  1. Your code does not quite work: if there is an interrupt from
>     another source while you are sleeping, you will read a wrong/old
>     value into your results
>
I don't think so. The waits are called until an random number is ready.

>  1. Why are you moving the power-ctrl into the read function instead
>     of calling them explicitly? This kind of breaks the concept of the
>     periph API...
>
Whats the benefit of this concept in this case? I have written tree
versions of this API wich is saving Power. This was the most simple and
efficient concept. From external view it makes no difference if RNG is
initialized only when needed or it has to power on and off manually.

**Comment by gebart** [comment]
Is it not better to defer power saving to the idle thread, and let other threads (if any exist) run while your thread is waiting for the RNG hardware?

Your current implementation will prevent any other threads from running since you force the hardware to switch to a power save mode.

Another solution would be to use `thread_yield()` and let any other threads run if they have something worthwhile to do while waiting for the RNG to be done with its thing.

**Comment by haukepetersen** [comment]
> I don't think so. The waits are called until an random number is ready.

Yap, you are right (I overlooked the loop...). Still, as @gebart suggests, the normal way to implement this in RIOT would be to put the current thread to sleep while waiting for results. But as the acquisition of new values is shorter then switching to another task (idle) and back I would say that your approach might be applicable here. 

Just a side question: why are you calling `_SEV` in between two `_WFE` calls instead of just calling `_WFE` once?

Regarding the power functions -> for this platform this might be working the way you propose, but RIOT supports many more, some of them behave differently and for them it makes sense to implement the power functions separately. Now application that use that peripheral should call those power functions anyway so they are portable...

**Comment by d00616** [comment]
> Just a side question: why are you calling |_SEV| in between two |_WFE|
> calls instead of just calling |_WFE| once?
>
I found this in Nordic developer forum. I had problems with sleep mode
with only one _WFE call. I check this later if it's possible to reduce
this to only one _WFE call. I'll describe in comments why this is changed.
>
> Regarding the power functions -> for this platform this might be
> working the way you propose, but RIOT supports many more, some of them
> behave differently and for them it makes sense to implement the power
> functions separately. Now application that use that peripheral should
> call those power functions anyway so they are portable...
>
It's comfortable to do this here. I agree this is not an perfect solution.

**Comment by d00616** [comment]
> Just a side question: why are you calling |_SEV| in between two |_WFE|
> calls instead of just calling |_WFE| once?
>
I have tested with calling _WFE one time. The while loop is executed
around 90 times. With two _WFE the loop is executed one time only. This
note looks right:
https://devzone.nordicsemi.com/question/1555/how-do-you-put-the-nrf51822-chip-to-sleep/

**Comment by d00616** [comment]
There is another problem with old random implementation. random_read doesn't work after powering off and on.

In my opinion my code can be merged.

**Comment by OlegHahm** [comment]
Travis disagrees:
```
Applying: cpu: nrf51822/periph/random Power optimization

cpu/nrf51822/periph/random.c:50: trailing whitespace.

+    

cpu/nrf51822/periph/random.c:50: trailing whitespace.

+    

ERROR: This change introduces new whitespace errors
```

**Comment by d00616** [comment]
Now its passing Travis checks.

File: cpu/nrf51822/periph/random.c [hunk 1]
```diff
@@ -25,33 +26,60 @@
 /* guard file in case no random device was specified */
 #if RANDOM_NUMOF
 
+uint8_t _rng_int = 0;
```
Comment (haukepetersen) [inline cpu/nrf51822/periph/random.c]: unused variable? Periph drivers should only allocate RAM if really really needed...

Comment (d00616) [inline cpu/nrf51822/periph/random.c]: Thanks. It's fixed now

**Comment by haukepetersen** [comment]
Ok, code works, Travis is happy (and so am I), then lets merge this: ACK and go.

**Pull request merged**